drivers/lora/sx1276.c

```diff
@@ -0,0 +1,717 @@
+		return -EIO;
+	}
+
+	ret = sx1276_write(dev, SX1276_REG_PAYLOAD_LENGTH, data_len);
+	if (ret < 0) {
+		LOG_ERR("Unable to write payload length");
+		return -EIO;
+	}
+
+	ret = sx1276_fifo_write(dev, data, data_len);
+	if (ret < 0)
+		return ret;
+
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_OPMODE_MODE_MASK;
+	regval |= SX1276_OPMODE_MODE_TX;
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sx1276_lora_config(struct device *dev,
+			      struct lora_modem_config *config)
+{
+	unsigned long long freq_rf;
+	int ret;
+	u8_t regval, pa_config;
+
+	/* Set sleep mode */
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, SX1276_OPMODE_MODE_SLEEP);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	/* Set LoRa mode */
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval |= SX1276_OPMODE_LONG_RANGE_MODE;
+	/* Set low frequency mode if required */
+	if (config->frequency < 525000000)
+		regval |= SX1276_OPMODE_LOW_FREQUENCY_MODE_ON;
+
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	/* Set standby mode */
+	ret = sx1276_read(dev, SX1276_REG_OPMODE, &regval, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read OPMODE");
+		return -EIO;
+	}
+
+	regval &= ~SX1276_OPMODE_MODE_MASK;
+	regval |= SX1276_OPMODE_MODE_STDBY;
+	ret = sx1276_write(dev, SX1276_REG_OPMODE, regval);
+	if (ret < 0) {
+		LOG_ERR("Unable to write OPMODE");
+		return -EIO;
+	}
+
+	k_sleep(200);
+
+	/* Set frequency */
+	freq_rf = config->frequency;
+	freq_rf *= (1 << 19);
+	freq_rf /= CLOCK_FREQ;
+
+	ret = sx1276_write(dev, SX1276_REG_FRF_MSB, (freq_rf >> 16 & 0xFF));
+	if (!ret)
+		ret = sx1276_write(dev, SX1276_REG_FRF_MID,
+				   (freq_rf >> 8 & 0xFF));
+	if (!ret)
+		ret = sx1276_write(dev, SX1276_REG_FRF_LSB, (freq_rf & 0xFF));
+	if (ret) {
+		LOG_ERR("Unable to write RF carrier frequency");
+		return -EIO;
+	}
+
+	/* Set output power */
+	ret = sx1276_read(dev, SX1276_REG_PA_CONFIG, &pa_config, 1);
+	if (ret < 0) {
+		LOG_ERR("Unable to read version PA config");
+		return -EIO;
+	}
+
+	pa_config &= ~GENMASK(3, 0);
+#if defined CONFIG_PA_RFO_PIN
+	pa_config &= ~SX1276_PA_CONFIG_PA_BOOST;
+	if (config->tx_power < sx1276_power[SX1276_PA_RFO].min)
+		config->tx_power = sx1276_power[SX1276_PA_RFO].min;
+	if (config->tx_power > sx1276_power[SX1276_PA_RFO].max)
+		config->tx_power = sx1276_power[SX1276_PA_RFO].max;
+#else
+	pa_config |= SX1276_PA_CONFIG_PA_BOOST;
+	if (config->tx_power > sx1276_power[SX1276_PA_BOOST].max - 3) {
+		if (config->tx_power > sx1276_power[SX1276_PA_BOOST].max)
+			config->tx_power = sx1276_power[SX1276_PA_BOOST].max;
+
+		config->tx_power -= 3;
```

Comment:
I'm not a LoRa expert, but is it safe to silently change a configuration that is passed by the user?